types/comparable_bytes.cc

```diff
@@ -36,21 +47,124 @@ static void convert_signed_fixed_length_integer(managed_bytes_view& src, bytes_o
+ * there are additional bytes in the encoding, followed by the two's complement of the number.
+ *
+ * (i.e) <n bits with inverted sign bit><2's complement of the number>
+ *       where n = number of bytes in encoding and n >= 1
+ *
+ * Because of the inverted sign bit, negative numbers compare smaller than positives, and because the length
+ * bits match the sign, longer positive numbers compare greater and longer negative ones compare smaller.
+ *
+ * Examples:
+ *      0              encodes as           80
+ *      1              encodes as           81
+ *     -1              encodes as           7F
+ *     63              encodes as           BF
+ *     64              encodes as           C040
+ *    -64              encodes as           40
+ *    -65              encodes as           3FBF
+ *   2^20-1            encodes as           EFFFFF
+ *   2^20              encodes as           F0100000
+ *  -2^20              encodes as           100000
+ *   2^63-1            encodes as           FFFFFFFFFFFFFFFFFF
+ *  -2^63              encodes as           000000000000000000
+ *
+ * The encoded value will have a maximum of 9 bytes including the length bits.
+ * As the number of bytes is specified in bits 2-9, no value is a prefix of another.
+ */
+static void encode_signed_long_type(int64_t value, bytes_ostream& out) {
+    // Create a negative mask : -1 for negative numbers and 0 for positive
+    int64_t negative_mask = value >> 63;
+    // If the value is negative, flip all bits so we can treat it like a positive number.
+    // This allows us to unify the transformation logic for both positive and negative values.
+    value ^= negative_mask;
+
+    // Count number of significant bits (1-63) that actually needed to be encoded.
+    // It can never be 64 as we have already flipped negative numbers.
+    int num_bits = 64 - std::countl_zero(uint64_t(value | 1));
+    // Calculate number of bytes required to represent the bits in comparable format.
+    // The encoding includes the encoded length as inverted sign bits and the
+    // actual value is stored as a 2's complement with atleast one sign bit preserved.
+    // For example, 32 (0b00100000) can be encoded in a single byte 0b10100000,
+    //          but 64 (0b01000000) requires two bytes : 0b00000001 0b01000000.
+    // So, 0-6 bits => 1 byte; 7-13 => 2 bytes; etc to 56-63 => 9 bytes.
+    // Note that 63 bits require 10 bytes but we handle it as a special case and
+    // encode it within 9 bytes - no sign bit is preserved and the decoding logic
+    // accounts for this scenario.
+    int num_bytes = std::min(9, num_bits / 7 + 1);
+    // Encode the number of bytes as inverted sign bits in the comparable format
+    std::array<char, 9> encoded_buffer;
+    if (num_bytes == 9) {
+        // An extra byte full of inverted sign bits has to be included in the comparable format
+        encoded_buffer[0] = negative_mask < 0 ? 0 : 0xFF;
```

Comment:
This is super minor but I can't resist: encoded_buffer[0] = uint8_t(negative_mask) ^ 0xff